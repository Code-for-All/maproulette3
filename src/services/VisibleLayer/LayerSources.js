import _isEmpty from 'lodash/isEmpty'
import _find from 'lodash/find'
import PropTypes from 'prop-types'
import messages from './Messages'

export const layerSourceShape = PropTypes.shape({
  /** Unique id for layer */
  layerId: PropTypes.string.isRequired,
  /** Human-readable name of layer formatted as react-intl message */
  name: PropTypes.object,
  /** Tile server URL. Supports substitutions: {x}, {y}, {z}, {accessToken} */
  url: PropTypes.string.isRequired,
  /** Human-readable attribution of layer formatted as react-intl message */
  attribution: PropTypes.object,
  /** Width/height of tiles */
  tileSize: PropTypes.number,
  /**
   * Set to true to adjust non-retina tiles for retina displays. tileSize may
   * also need to be adjusted.
   *
   * @see See http://leafletjs.com/reference-1.0.3.html#tilelayer-detectretina
   */
  detectRetina: PropTypes.bool,
  /** Access token for substitution in tile server url */
  accessToken: PropTypes.string,
  /** Set to true to mark as the default layer */
  default: PropTypes.bool,
  /** Set to false if layer should not be offered to users for selection */
  isSelectable: PropTypes.bool,
  /** Set to true for dynamically-created layers, such as custom basemaps */
  isDynamic: PropTypes.bool,
})

export const OPEN_STREET_MAP = 'OpenStreetMap'
export const OPEN_CYCLE_MAP = 'OpenCycleMap'
export const BING = 'Bing'
export const MAPBOX_STREETS = 'Mapbox'
export const MAPBOX_LIGHT = 'MapboxLight'
export const MAPBOX_SATELLITE_STREETS = 'MapboxSatellite'

/** Array of available layer sources */
export const LayerSources = [{
    layerId: OPEN_STREET_MAP,
    name: messages.openStreetMapName,
    url: process.env.REACT_APP_OPEN_STREET_MAP_TILESERVER_URL,
    attribution: messages.openStreetMapAttribution,
  }, {
    layerId: OPEN_CYCLE_MAP,
    name: messages.openCycleMapName,
    url: process.env.REACT_APP_OPEN_CYCLE_MAP_TILESERVER_URL,
    attribution: messages.openCycleMapAttribution,
  }, {
    layerId: BING,
    name: messages.bingName,
    url: process.env.REACT_APP_BING_MAP_TILESERVER_URL,
    attribution: messages.bingAttribution,
    detectRetina: true,
    tileSize: 256,
  }]

// If a Mapbox access/api token has been provided, then add Mapbox layer
// sources.
if (!_isEmpty(process.env.REACT_APP_MAPBOX_ACCESS_TOKEN)) {
  LayerSources.push({
    layerId: MAPBOX_STREETS,
    name: messages.mapboxStreetsName,
    url: process.env.REACT_APP_MAPBOX_STREETS_MAP_TILESERVER_URL,
    accessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN,
    attribution: messages.mapboxAttribution,
    default: true,
  })

  LayerSources.push({
    layerId: MAPBOX_SATELLITE_STREETS,
    name: messages.mapboxSatelliteName,
    url: process.env.REACT_APP_MAPBOX_SATELLITE_STREETS_MAP_TILESERVER_URL,
    accessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN,
    attribution: messages.mapboxAttribution,
  })

  LayerSources.push({
    layerId: MAPBOX_LIGHT,
    name: messages.mapboxLightName,
    url: process.env.REACT_APP_MAPBOX_LIGHT_MAP_TILESERVER_URL,
    accessToken: process.env.REACT_APP_MAPBOX_ACCESS_TOKEN,
    attribution: messages.mapboxAttribution,
  })
}

/**
 * Returns a default layer source for use in situations where no layer source
 * has been specified.
 */
export const defaultLayerSource = function() {
  const configuredDefault = _find(LayerSources, {default: true})
  return configuredDefault ? configuredDefault : LayerSources[0]
}

/**
 * Retrieves the (static) layer source with a matching id. Dynamic layer
 * sources are not searched.
 */
export const layerSourceWithId = function(layerId) {
  return _find(LayerSources, {layerId})
}

/**
 * Create and return a dynamic layer source with the given layerId
 * and url. Primarily intended for use with custom basemaps.
 */
export const createDynamicLayerSource = function(layerId, url) {
  return {
    layerId,
    name: messages.customName,
    url,
    isSelectable: false,
    isDynamic: true,
  }
}
