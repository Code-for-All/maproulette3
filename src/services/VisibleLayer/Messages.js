import { defineMessages } from 'react-intl'

/**
 * Internationalized messages for use with VisibleLayer
 */
export default defineMessages({
  openStreetMapName: {
    id: "map.name.openStreetMap",
    defaultMessage: "OpenStreetMap",
  },

  openStreetMapAttribution: {
    id: "map.attribution.openStreetMap",
    defaultMessage: "&copy; <a href='http://openstreetmap.org'>OpenStreetMap</a> contributors, <a href='http://creativecommons.org/licenses/by-sa/2.0/'>CC-BY-SA</a>",
  },

  openCycleMapName: {
    id: "map.name.openCycleMap",
    defaultMessage: "OpenCycleMap",
  },

  openCycleMapAttribution: {
    id: "map.attribution.openCycleMap",
    defaultMessage: "&copy; OpenCycleMap, Map data &copy; <a href='http://openstreetmap.org'>OpenStreetMap</a> contributors, <a href='http://creativecommons.org/licenses/by-sa/2.0/'>CC-BY-SA</a>",
  },

  bingName: {
    id: "map.name.bing",
    defaultMessage: "Bing",
  },

  bingAttribution: {
    id: "map.attribution.bing",
    defaultMessage: "Tiles &copy; Esri &mdash; Source: Esri, i-cubed, USDA, USGS, AEX, GeoEye, Getmapping, Aerogrid, IGN, IGP, UPR-EGP, and the GIS User Community",
  },

  mapboxStreetsName: {
    id: "map.name.mapbox",
    defaultMessage: "Mapbox",
  },

  mapboxLightName: {
    id: "map.name.mapboxLight",
    defaultMessage: "Mapbox Light",
  },

  mapboxSatelliteName: {
    id: "map.name.mapboxSatellite",
    defaultMessage: "Mapbox Satellite",
  },

  mapboxAttribution: {
    id: "map.attribution.mapbox",
    defaultMessage: "&copy <a href='https://www.mapbox.com/about/maps/'>Mapbox</a>, &copy <a href='http://www.openstreetmap.org/copyright'>OpenStreetMap</a> <strong><a href='https://www.mapbox.com/map-feedback/' target='_blank'>Improve this map</a></strong>",
  },

  customName: {
    id: "map.name.custom",
    defaultMessage: "Custom",
  }
})
