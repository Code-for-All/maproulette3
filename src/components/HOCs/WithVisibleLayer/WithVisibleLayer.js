import { connect } from 'react-redux'
import _get from 'lodash/get'
import _isNumber from 'lodash/isNumber'
import _isString from 'lodash/isString'
import _isObject from 'lodash/isObject'
import _isEmpty from 'lodash/isEmpty'
import { layerSourceWithId,
         defaultLayerSource,
         createDynamicLayerSource }
       from '../../../services/VisibleLayer/LayerSources'
import { changeVisibleLayer } from '../../../services/VisibleLayer/VisibleLayer'
import { ChallengeBasemap,
         BasemapLayerSources }
       from '../../../services/Challenge/ChallengeBasemap/ChallengeBasemap'
import WithCurrentUser from '../WithCurrentUser/WithCurrentUser'
import WithChallengePreferences
       from '../WithChallengePreferences/WithChallengePreferences'

/**
 * WithVisibleLayer provides the wrapped component with the proper tile layer
 * source to be displayed in a map based on the current state of the redux
 * store or -- if there is no current active layer -- the configuration of the
 * layer sources themselves. It also provides a function for changing the active
 * layer.
 *
 * @see See LayerSources
 *
 * @author [Neil Rotstan](https://github.com/nrotstan)
 */
const WithVisibleLayer = WrappedComponent =>
  WithCurrentUser(WithChallengePreferences(
    connect(mapStateToProps, mapDispatchToProps)(WrappedComponent)
  ))

/**
 * Determine which map layer should be used as a default. If a challenge
 * default basemap is given, it will be used. Otherwise if a user default
 * basemap is given, it will be used. Otherwise if a `defaultLayer` prop was
 * given, it will be used. Otherwise the default layer source from LayerSources
 * will be used.
 */
export const defaultLayer = ownProps => {
  const challengeDefaultBasemap = _get(ownProps, 'challenge.defaultBasemap')
  const userDefaultBasemap = _get(ownProps, 'user.settings.defaultBasemap')
  let layer = null

  if (_isNumber(challengeDefaultBasemap) &&
      challengeDefaultBasemap !== ChallengeBasemap.none) {
    if (challengeDefaultBasemap !== ChallengeBasemap.custom) {
      layer = layerSourceWithId(BasemapLayerSources[challengeDefaultBasemap])
    }
    else if (!_isEmpty(ownProps.challenge.customBasemap)) {
      layer = createDynamicLayerSource(`challenge_${ownProps.challenge.id}`,
                                       ownProps.challenge.customBasemap)
    }
  }
  else if (_isNumber(userDefaultBasemap) &&
           userDefaultBasemap !== ChallengeBasemap.none) {
    if (userDefaultBasemap !== ChallengeBasemap.custom) {
      layer = layerSourceWithId(BasemapLayerSources[userDefaultBasemap])
    }
    else if (!_isEmpty(ownProps.user.settings.customBasemap)) {
      layer = createDynamicLayerSource(`user_${ownProps.user.id}`,
                                       ownProps.user.settings.customBasemap)
    }
  }
  else if (_isObject(ownProps.defaultLayer)) {
    layer = ownProps.defaultLayer
  }
  else if (_isString(ownProps.defaultLayer)) {
    layer = layerSourceWithId(ownProps.defaultLayer)
  }

  return layer ? layer : defaultLayerSource()
}

export const mapStateToProps = (state, ownProps) => {
  const challengeId = _get(ownProps, 'challenge.id')
  let source = null

  if (_isNumber(challengeId)) {
    if (_isString(ownProps.visibleMapLayer)) {
      source = layerSourceWithId(ownProps.visibleMapLayer)
    }
  }
  else if (state.visibleLayer) {
    source = state.visibleLayer
  }

  return {
    source: source ? source : defaultLayer(ownProps)
  }
}

export const mapDispatchToProps = (dispatch, ownProps) => {
  return {
    changeLayer: layerId => {
      const isVirtual = _isNumber(ownProps.virtualChallengeId)
      const challengeId = isVirtual ? ownProps.virtualChallengeId :
                                      _get(ownProps, 'challenge.id')

      if (_isNumber(challengeId) && ownProps.setVisibleMapLayer) {
        ownProps.setVisibleMapLayer(challengeId, isVirtual, layerId)
      }
      else {
        dispatch(changeVisibleLayer(layerId))
      }
    },
  }
}

export default WithVisibleLayer
