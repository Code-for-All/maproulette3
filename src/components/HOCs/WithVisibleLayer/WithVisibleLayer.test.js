import React, { Component } from 'react'
import { mapStateToProps, mapDispatchToProps } from './WithVisibleLayer'
import { layerSourceWithName,
         defaultLayerSource } from '../../../services/VisibleLayer/LayerSources'
import { changeVisibleLayer } from '../../../services/VisibleLayer/VisibleLayer'

jest.mock('../../../services/VisibleLayer/LayerSources')
jest.mock('../../../services/VisibleLayer/VisibleLayer')

layerSourceWithName.mockImplementation((layer) => layer)

let basicState = null

beforeEach(() => {
  basicState = {}
})

test("mapStateToProps maps source to visibleLayer from state, if any", () => {
  basicState = {visibleLayer: "layer1"}
  const mappedProps = mapStateToProps(basicState, {defaultLayer: "layer2"})
  expect(mappedProps.source).toEqual(basicState.visibleLayer)

  expect(mappedProps).toMatchSnapshot()
})

test("mapStateToProps source uses given default layer if no visible layer in state", () => {
  const mappedProps = mapStateToProps({}, {defaultLayer: "layer2"})
  expect(mappedProps.source).toEqual("layer2")

  expect(mappedProps).toMatchSnapshot()
})

test("mapStateToProps source uses call to defaultLayerSource() if neither default layer nor visible layer is available", () => {
  const mappedProps = mapStateToProps({}, {})

  expect(defaultLayerSource).toBeCalled()

  expect(mappedProps).toMatchSnapshot()
})

test("mapDispatchToProps maps function changeLayer", () => {
  const dispatch = jest.fn(() => Promise.resolve())
  const mappedProps = mapDispatchToProps(dispatch, {})

  mappedProps.changeLayer("myLayer")
  expect(dispatch).toBeCalled()
  expect(changeVisibleLayer).toBeCalledWith("myLayer")
})
