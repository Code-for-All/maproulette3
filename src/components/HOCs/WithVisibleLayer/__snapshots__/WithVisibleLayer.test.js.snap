// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mapStateToProps maps source to visibleLayer from state, if any 1`] = `
Object {
  "source": "layer1",
}
`;

exports[`mapStateToProps source uses call to defaultLayerSource() if neither default layer nor visible layer is available 1`] = `
Object {
  "source": undefined,
}
`;

exports[`mapStateToProps source uses given default layer if no visible layer in state 1`] = `
Object {
  "source": "layer2",
}
`;
